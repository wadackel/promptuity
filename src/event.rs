//! A module to read events.
//!
//! # Event
//!
//! The `event` module provides functionalities used in prompt handlers.
//! The main offering is the direct re-export of [`crossterm::event`].
//!
//! ## Examples
//!
//! ```ignore
//! use promptuity::event::{KeyCode, KeyModifiers};
//!
//! impl<W: std::io::Write> Prompt<W> for YourPrompt {
//!   // ...
//!   fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
//!     match (code, modifiers) {
//!       (KeyCode::Enter, KeyModifiers::NONE) => PromptState::Submit,
//!       _ => PromptState::Active,
//!     }
//!   }
//!   // ...
//! }
//! ```

/// [`KeyCode`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyCode;
/// [`KeyModifiers`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyModifiers;
