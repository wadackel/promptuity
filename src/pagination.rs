//! A module that assists with pagination functionality.
//!
//! # Pagination
//!
//! The `pagination` module provides functionalities to assist with pagination.  
//! It offers generic logic as a utility, so it can be expected to be useful in your custom prompts as well.
//!
//! In built-in prompts, it is used in [`crate::prompts::Select`] and [`crate::prompts::MultiSelect`].
//!
//! ## Examples
//!
//! ```
//! use promptuity::pagination::{Page, paginate};
//!
//! let page_size = 5;
//! let items = vec![1, 2, 3, 4, 5, 6, 7, 8];
//!
//! assert_eq!(paginate(page_size, &items, 1), Page {
//!     first: true,
//!     last: false,
//!     items: &[1, 2, 3, 4, 5],
//!     cursor: 1,
//!     total: 8,
//! });
//!
//! assert_eq!(paginate(page_size, &items, 3), Page {
//!     first: false,
//!     last: false,
//!     items: &[2, 3, 4, 5, 6],
//!     cursor: 2,
//!     total: 8,
//! });
//!
//! assert_eq!(paginate(page_size, &items, 7), Page {
//!     first: false,
//!     last: true,
//!     items: &[4, 5, 6, 7, 8],
//!     cursor: 4,
//!     total: 8,
//! });
//! ```

/// A page of items.
#[derive(Debug, PartialEq)]
pub struct Page<'a, T> {
    /// A flag indicating whether this is the first page.
    pub first: bool,
    /// A flag indicating whether this is the last page.
    pub last: bool,
    /// The items in this page.
    pub items: &'a [T],
    /// The cursor position in this page.
    pub cursor: usize,
    /// The total number of items.
    pub total: usize,
}

/// Paginates the given items.
pub fn paginate<T>(size: usize, items: &[T], current: usize) -> Page<'_, T> {
    let (begin, end, cursor) = if items.len() <= size {
        (0, items.len(), current)
    } else if current < size / 2 {
        (0, size, current)
    } else if items.len() - current - 1 < size / 2 {
        let begin = items.len() - size;
        (begin, items.len(), current - begin)
    } else {
        (current - size / 2, current + (size - size / 2), size / 2)
    };

    Page {
        first: begin == 0,
        last: end == items.len(),
        items: &items[begin..end],
        cursor,
        total: items.len(),
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    fn range(r: std::ops::Range<usize>) -> Vec<usize> {
        (r.start..r.end + 1).collect()
    }

    #[test]
    fn test_page() {
        let items_1_2 = range(1..2);
        let items_1_5 = range(1..5);
        let items_4_8 = range(4..8);
        let items_9_13 = range(9..13);
        let items_11_15 = range(11..15);

        let tests = vec![
            (
                5,
                range(1..15),
                0,
                Page {
                    first: true,
                    last: false,
                    items: &items_1_5,
                    cursor: 0,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                2,
                Page {
                    first: true,
                    last: false,
                    items: &items_1_5,
                    cursor: 2,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                5,
                Page {
                    first: false,
                    last: false,
                    items: &items_4_8,
                    cursor: 2,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                10,
                Page {
                    first: false,
                    last: false,
                    items: &items_9_13,
                    cursor: 2,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                12,
                Page {
                    first: false,
                    last: true,
                    items: &items_11_15,
                    cursor: 2,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                13,
                Page {
                    first: false,
                    last: true,
                    items: &items_11_15,
                    cursor: 3,
                    total: 15,
                },
            ),
            (
                5,
                range(1..15),
                14,
                Page {
                    first: false,
                    last: true,
                    items: &items_11_15,
                    cursor: 4,
                    total: 15,
                },
            ),
            (
                5,
                range(1..5),
                3,
                Page {
                    first: true,
                    last: true,
                    items: &items_1_5,
                    cursor: 3,
                    total: 5,
                },
            ),
            (
                5,
                range(1..2),
                1,
                Page {
                    first: true,
                    last: true,
                    items: &items_1_2,
                    cursor: 1,
                    total: 2,
                },
            ),
        ];
        for (size, items, current, expected) in tests {
            assert_eq!(expected, paginate(size, &items, current));
        }
    }
}
