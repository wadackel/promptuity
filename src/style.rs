//! A module to apply attributes and colors on your text.
//!
//! # Style
//!
//! The `style` module offers functionality to apply styling to your text.
//! In promptuity, it exposes a simple API that wraps the features of [`crossterm::style`].
//!
//! ## Examples
//!
//! This is an example of styling text.
//!
//! ```
//! use promptuity::style::{Color, Styled};
//!
//! let styled_text = Styled::new("Colored text").fg(Color::Green).bold().to_string();
//!
//! assert_eq!(styled_text, "\u{1b}[38;5;10m\u{1b}[1mColored text\u{1b}[0m");
//! ```
//!
//! This is an example of changing the displayed text based on whether Unicode is supported or not.  
//! Useful for representing icons and other symbols.
//!
//! ```no_run
//! use promptuity::style::Symbol;
//!
//! //              Unicode (main)       ASCII (fallback)
//! //                            \     /
//! let unselected_icon = Symbol("◯", "[ ]");
//! let selected_icon = Symbol("◉", "[x]");
//! ```

/// [`Color`] re-exports from [`crossterm::style::Color`].
pub use crossterm::style::Color;
use crossterm::style::{Attribute, Attributes, ContentStyle, Stylize};

/// A styling utility for strings wrapped in [`crossterm::style::ContentStyle`].
#[derive(Debug)]
pub struct Styled {
    content: String,
    fg: Option<Color>,
    bg: Option<Color>,
    attr: Option<Attributes>,
}

impl Styled {
    pub fn new(content: impl std::fmt::Display) -> Self {
        Self {
            content: content.to_string(),
            fg: None,
            bg: None,
            attr: None,
        }
    }

    pub fn fg(&mut self, color: Color) -> &mut Self {
        self.fg = Some(color);
        self
    }

    pub fn bg(&mut self, color: Color) -> &mut Self {
        self.bg = Some(color);
        self
    }

    pub fn bold(&mut self) -> &mut Self {
        self.attr(Attribute::Bold)
    }

    pub fn italic(&mut self) -> &mut Self {
        self.attr(Attribute::Italic)
    }

    pub fn underline(&mut self) -> &mut Self {
        self.attr(Attribute::Underlined)
    }

    pub fn dim(&mut self) -> &mut Self {
        self.attr(Attribute::Dim)
    }

    pub fn rev(&mut self) -> &mut Self {
        self.attr(Attribute::Reverse)
    }

    pub fn blink(&mut self) -> &mut Self {
        self.attr(Attribute::SlowBlink)
    }

    fn attr(&mut self, attr: Attribute) -> &mut Self {
        match self.attr {
            Some(ref mut attrs) => attrs.set(attr),
            None => self.attr = Some(Attributes::from(attr)),
        }
        self
    }
}

impl std::fmt::Display for Styled {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut style = ContentStyle::new();
        if let Some(fg) = self.fg {
            style = style.with(fg);
        }
        if let Some(bg) = self.bg {
            style = style.on(bg);
        }
        if let Some(attr) = &self.attr {
            style.attributes = *attr;
        }
        write!(f, "{}", style.apply(self.content.as_str()))
    }
}

#[cfg(windows)]
fn is_unicode_supported() -> bool {
    use std::env;

    if env::var("WT_SESSION").is_ok() {
        return true;
    }

    if let Ok(program) = env::var("TERM_PROGRAM") {
        match program.as_str() {
            "vscode" => return true,
            _ => {}
        }
    }

    if let Ok(term) = env::var("TERM") {
        match term.as_str() {
            "xterm-256color" | "alacritty" => return true,
            _ => {}
        }
    }

    false
}

#[cfg(unix)]
fn is_unicode_supported() -> bool {
    true
}

/// A symbol that can be displayed in unicode or ascii.
#[derive(Debug, Copy, Clone)]
pub struct Symbol<'a, 'b>(
    /// The main symbol to display when unicode is supported.
    pub &'a str,
    /// The fallback symbol to display when unicode is not supported.
    pub &'b str,
);

impl<'a, 'b> Symbol<'a, 'b> {
    /// Creates a new [`Symbol`].
    pub fn new(main: &'a str, fallback: &'b str) -> Self {
        Self(main, fallback)
    }
}

impl<'a, 'b> std::fmt::Display for Symbol<'a, 'b> {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if is_unicode_supported() {
            write!(f, "{}", self.0)
        } else {
            write!(f, "{}", self.1)
        }
    }
}
