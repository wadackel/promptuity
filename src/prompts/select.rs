use crate::event::*;
use crate::pagination::paginate;
use crate::style::*;
use crate::{Error, Prompt, PromptBody, PromptInput, PromptState, RenderPayload};

const S_UNSELECTED: Symbol = Symbol("◯", "[ ]");
const S_SELECTED: Symbol = Symbol("◉", "[x]");

/// A struct representing an option in the [`Select`] prompt.
#[derive(Debug, Clone)]
pub struct SelectOption<T: Default + Clone> {
    /// The label of the option.
    pub label: String,
    /// The value of the option.
    pub value: T,
    /// The hint message of the option. Defaults to `None`.
    pub hint: Option<String>,
}

impl<T: Default + Clone> SelectOption<T> {
    /// Creates a new [`SelectOption`] with the given label and value.
    pub fn new(label: impl std::fmt::Display, value: T) -> Self {
        Self {
            label: label.to_string(),
            value,
            hint: None,
        }
    }

    /// Sets the hint message for the option.
    pub fn with_hint(mut self, hint: impl std::fmt::Display) -> Self {
        self.hint = Some(hint.to_string());
        self
    }
}

/// A trait for customizing the display of [`Select`].
///
/// All methods have default implementations, allowing you to override only the specific formatting process you need.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Select, SelectOption, SelectFormatter};
///
/// struct CustomFormatter;
///
/// impl SelectFormatter for CustomFormatter {
///     fn option_icon(&self, active: bool) -> String {
///        if active {
///            ">".into()
///        } else {
///            " ".into()
///        }
///     }
/// }
///
/// let _ = Select::new("...", vec![SelectOption::new("...", "...")]).with_formatter(CustomFormatter);
/// ```
pub trait SelectFormatter {
    /// Icons displayed for each option.
    fn option_icon(&self, active: bool) -> String {
        if active {
            Styled::new(S_SELECTED).fg(Color::Green).to_string()
        } else {
            Styled::new(S_UNSELECTED).fg(Color::DarkGrey).to_string()
        }
    }

    /// Formats the label of the option.
    fn option_label(&self, label: String, active: bool) -> String {
        if active {
            Styled::new(label).underline().to_string()
        } else {
            Styled::new(label).fg(Color::DarkGrey).to_string()
        }
    }

    /// Formats the hint message of the option.
    fn option_hint(&self, hint: Option<String>, active: bool) -> String {
        let _ = active;
        hint.as_ref().map_or_else(String::new, |hint| {
            format!(
                " {}",
                Styled::new(format!("({})", hint)).fg(Color::DarkGrey)
            )
        })
    }

    /// Formats the option.
    fn option(&self, icon: String, label: String, hint: String, active: bool) -> String {
        let _ = active;
        format!("{} {}{}", icon, label, hint)
    }
}

/// The default formatter for [`Select`].
pub struct DefaultSelectFormatter;

impl DefaultSelectFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self {}
    }
}

impl SelectFormatter for DefaultSelectFormatter {}

/// A prompt for selecting a single element from a list of options.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`SelectFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Page Size**: The total number of options to displayed per page, used for pagination. Defaults to `8`.
///
/// # Notes
///
/// Passing an empty `options` will result in an error. Please ensure to provide `options` with at least one item.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Select, SelectOption};
///
/// let _ = Select::new("What is your favorite color?", vec![
///     SelectOption::new("Red", "#ff0000"),
///     SelectOption::new("Green", "#00ff00").with_hint("recommended"),
///     SelectOption::new("Blue", "#0000ff"),
/// ]).with_page_size(5);
/// ```
pub struct Select<T: Default + Clone> {
    formatter: Box<dyn SelectFormatter>,
    message: String,
    hint: Option<String>,
    page_size: usize,
    options: Vec<SelectOption<T>>,
    index: usize,
}

impl<T: Default + Clone> Select<T> {
    /// Creates a new [`Select`] prompt with the given message and options.
    pub fn new(message: impl std::fmt::Display, options: Vec<SelectOption<T>>) -> Self {
        Self {
            formatter: Box::new(DefaultSelectFormatter::new()),
            message: message.to_string(),
            hint: None,
            page_size: 8,
            options,
            index: 0,
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl SelectFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the page size for the prompt.
    pub fn with_page_size(&mut self, page_size: usize) -> &mut Self {
        self.page_size = page_size;
        self
    }
}

impl<T: Default + Clone> AsMut<Select<T>> for Select<T> {
    fn as_mut(&mut self) -> &mut Select<T> {
        self
    }
}

impl<T: Default + Clone> Prompt for Select<T> {
    type Output = T;

    fn setup(&mut self) -> Result<(), Error> {
        if self.options.is_empty() {
            return Err(Error::Config("options cannot be empty.".into()));
        }
        Ok(())
    }

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> crate::PromptState {
        match (code, modifiers) {
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Enter, _) | (KeyCode::Char(' '), _) => PromptState::Submit,
            (KeyCode::Up, _)
            | (KeyCode::Char('k'), _)
            | (KeyCode::Char('p'), KeyModifiers::CONTROL) => {
                self.index = self.index.saturating_sub(1);
                PromptState::Active
            }
            (KeyCode::Down, _)
            | (KeyCode::Char('j'), _)
            | (KeyCode::Char('n'), KeyModifiers::CONTROL) => {
                self.index = std::cmp::min(
                    self.options.len().saturating_sub(1),
                    self.index.saturating_add(1),
                );
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn submit(&mut self) -> Self::Output {
        let option = self.options.get(self.index).unwrap();
        option.value.clone()
    }

    fn render(&mut self, state: &PromptState) -> Result<RenderPayload, String> {
        let payload = RenderPayload::new(self.message.clone(), self.hint.clone(), None);

        match state {
            PromptState::Submit => {
                let option = self.options.get(self.index).unwrap();
                Ok(payload.input(PromptInput::Raw(option.label.clone())))
            }

            _ => {
                let page = paginate(self.page_size, &self.options, self.index);
                let options = page
                    .items
                    .iter()
                    .enumerate()
                    .map(|(i, option)| {
                        let active = i == page.cursor;
                        self.formatter.option(
                            self.formatter.option_icon(active),
                            self.formatter.option_label(option.label.clone(), active),
                            self.formatter.option_hint(option.hint.clone(), active),
                            active,
                        )
                    })
                    .collect::<Vec<_>>()
                    .join("\n");

                let raw = options.to_string();

                Ok(payload.body(PromptBody::Raw(raw)))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    macro_rules! options {
        ($count: expr) => {{
            let mut options = Vec::new();
            for i in 1..=$count {
                options.push(SelectOption::new(
                    format!("Value{}", i),
                    format!("value{}", i),
                ));
            }
            options
        }};
    }

    test_prompt!(
        test_hint,
        Select::new("test message", options!(3)).with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_10_items,
        Select::new("test message", options!(10)).as_mut(),
        vec![]
    );

    test_prompt!(
        test_10_items_with_5_page_size,
        Select::new("test message", options!(10)).with_page_size(5),
        vec![]
    );

    test_prompt!(
        test_option_hint,
        Select::new(
            "test message",
            vec![
                SelectOption::new("Value1", "value1".to_string()).with_hint("hint1"),
                SelectOption::new("Value2", "value2".to_string()),
                SelectOption::new("Value3", "value3".to_string()).with_hint("hint3"),
            ]
        )
        .with_page_size(5),
        vec![]
    );

    test_prompt!(
        test_move,
        Select::new("test message", options!(10)).with_page_size(5),
        vec![
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Char('n'), KeyModifiers::CONTROL),
            (KeyCode::Char('k'), KeyModifiers::NONE),
            (KeyCode::Char('p'), KeyModifiers::CONTROL),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE), // 10
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE), // 1
            (KeyCode::Up, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_select_5,
        Select::new("test message", options!(10)).as_mut(),
        vec![
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );
}
