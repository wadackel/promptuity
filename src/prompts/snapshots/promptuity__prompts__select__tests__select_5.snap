---
source: src/prompts/select.rs
---
state: Active
message: test message
hint: none
placeholder: none
input(none):
body(raw):
[38;5;10mâ—‰[39m [4mValue1[0m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
[38;5;8mâ—¯[39m [38;5;8mValue6[39m
[38;5;8mâ—¯[39m [38;5;8mValue7[39m
[38;5;8mâ—¯[39m [38;5;8mValue8[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;10mâ—‰[39m [4mValue2[0m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
[38;5;8mâ—¯[39m [38;5;8mValue6[39m
[38;5;8mâ—¯[39m [38;5;8mValue7[39m
[38;5;8mâ—¯[39m [38;5;8mValue8[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;10mâ—‰[39m [4mValue3[0m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
[38;5;8mâ—¯[39m [38;5;8mValue6[39m
[38;5;8mâ—¯[39m [38;5;8mValue7[39m
[38;5;8mâ—¯[39m [38;5;8mValue8[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;10mâ—‰[39m [4mValue4[0m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
[38;5;8mâ—¯[39m [38;5;8mValue6[39m
[38;5;8mâ—¯[39m [38;5;8mValue7[39m
[38;5;8mâ—¯[39m [38;5;8mValue8[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;10mâ—‰[39m [4mValue5[0m
[38;5;8mâ—¯[39m [38;5;8mValue6[39m
[38;5;8mâ—¯[39m [38;5;8mValue7[39m
[38;5;8mâ—¯[39m [38;5;8mValue8[39m
---
state: Submit
input(raw): Value5
body(none):
